GVFS/GVFS.Common/NuGetUpgrade/OrgNuGetUpgrader.cs

```diff
@@ -0,0 +1,244 @@
+using System.Threading.Tasks;
+
+namespace GVFS.Common.NuGetUpgrade
+{
+    public class OrgNuGetUpgrader : NuGetUpgrader
+    {
+        private HttpClient httpClient;
+        private string platform;
+
+        public OrgNuGetUpgrader(
+           string currentVersion,
+           ITracer tracer,
+           PhysicalFileSystem fileSystem,
+           HttpClient httpClient,
+           bool dryRun,
+           bool noVerify,
+           OrgNuGetUpgraderConfig config,
+           string downloadFolder,
+           string platform,
+           ICredentialStore credentialStore)
+           : base(
+               currentVersion,
+               tracer,
+               fileSystem,
+               dryRun,
+               noVerify,
+               config,
+               downloadFolder,
+               credentialStore)
+        {
+            this.httpClient = httpClient;
+            this.platform = platform;
+        }
+
+        public OrgNuGetUpgrader(
+           string currentVersion,
+           ITracer tracer,
+           PhysicalFileSystem fileSystem,
+           HttpClient httpClient,
+           bool dryRun,
+           bool noVerify,
+           OrgNuGetUpgraderConfig config,
+           string platform,
+           NuGetFeed nuGetFeed,
+           ICredentialStore credentialStore)
+           : base(
+               currentVersion,
+               tracer,
+               dryRun,
+               noVerify,
+               fileSystem,
+               config,
+               nuGetFeed,
+               credentialStore)
+        {
+            this.httpClient = httpClient;
+            this.platform = platform;
+        }
+
+        public override bool SupportsAnonymousVersionQuery { get => true; }
+
+        private OrgNuGetUpgraderConfig Config { get => this.nuGetUpgraderConfig as OrgNuGetUpgraderConfig;  }
+        private string OrgInfoServerUrl { get => this.Config.OrgInfoServer; }
+        private string Ring { get => this.Config.UpgradeRing; }
+
+        public static bool TryCreate(
+            ITracer tracer,
+            PhysicalFileSystem fileSystem,
+            LocalGVFSConfig gvfsConfig,
+            HttpClient httpClient,
+            ICredentialStore credentialStore,
+            bool dryRun,
+            bool noVerify,
+            out OrgNuGetUpgrader upgrader,
+            out string error)
+        {
+            OrgNuGetUpgraderConfig upgraderConfig = new OrgNuGetUpgraderConfig(tracer, gvfsConfig);
+            upgrader = null;
+
+            if (!upgraderConfig.TryLoad(out error))
+            {
+                upgrader = null;
+                return false;
+            }
+
+            if (!upgraderConfig.IsConfigured(out error))
+            {
+                return false;
+            }
+
+            if (!upgraderConfig.IsReady(out error))
+            {
+                return false;
+            }
+
+            string platform = RuntimeInformation.IsOSPlatform(OSPlatform.OSX) ? "macOS" : "windows";

```

Comment:
We've tried to avoid adding calls to `RuntimeInformation.IsOSPlatform` in VFS4G.  Rather than calling `IsOSPlatform` could you add a new `GetPlatformName` type method to `GVFSPlatform`?

Reply:
> Rather than calling IsOSPlatform could you add a new GetPlatformName type method to GVFSPlatform?

I can move this logic, however - my concern is that `GVFSPlatform` introduces some questions with testing. Ideally, I would like to be able to test with different platforms. Is that possible with the static GVFSPlatform instance?

Reply:
Currently the unit tests register `MockPlatform` as the active `GVFSPlatform`:

```
    [SetUpFixture]
    public class Setup
    {
        [OneTimeSetUp]
        public void SetUp()
        {
            GVFSPlatform.Register(new MockPlatform());
        }
    }
```

And so you should be able to add a `SetCurrentPlatform` type method to `MockPlatform` to allow you to set different platforms for testing.

Reply:
(This code is in GVFS/GVFS.UnitTests/Setup.cs)

As an alternative/enhancement we could add another method to `GVFSPlatform` that acts like `Register` but can be called multiple times (for unit testing only) so that we could swap in different mocks.

Reply:
> And so you should be able to add a SetCurrentPlatform type method to MockPlatform to allow you to set different platforms for testing.

Does this get cleared / reset between each test? or wpuld I run the risk of polluting / affecting other tests?

Reply:
Based on this:

> The OneTimeSetUp method in a SetUpFixture is executed once before any of the fixtures contained in its namespace. The OneTimeTearDown method is executed once after all the fixtures have completed execution.

https://github.com/nunit/docs/wiki/SetUpFixture-Attribute

I would expect it's a one time thing.  However, until we have something like this:

> As an alternative/enhancement we could add another method to GVFSPlatform that acts like Register but can be called multiple times (for unit testing only) so that we could swap in different mocks.

We could use a per-test Teardown in the fixture(s) that call `SetCurrentPlatform` to ensure we leave the platform name with its expected value.

Reply:
I think longer term we should revisit this pattern, as it makes it harder to test, and consider alternate patterns that are easier to test. For now, I went ahead and added properties off of the GVFS platform.

Reply:
> I think longer term we should revisit this pattern, as it makes it harder to test, and consider alternate patterns that are easier to test. For now, I went ahead and added properties off of the GVFS platform.

I'm happy to do some brainstorming in this area with you for a future PR.  

If there is a way we can keep runtime platform checks out of the VFS4G code and make the code easier to test that sounds like a win.